feat: add FOCA boilerplate

JaeAeich opened the pull request:
- Add [FOCA](https://github.com/elixir-cloud-aai/foca) archetype configuration for API server
- Add dummy API controllers
- API server starts up with (non-functional) Swagger UI at `http://127.0.0.1:8080/ga4gh/tes/v1/ui`; API specs hosted at `http://127.0.0.1:8080/ga4gh/tes/v1/openapi.json`

Fixes #156

uniqueg commented on tesk/__init__.py:
See comment in #177 regarding package-level docstrings.

uniqueg commented on tesk/api/task_execution_service.openapi.yaml:
Adopt the directory structure as discussed in #156. Goes for this and other API-related files.

uniqueg commented on tesk/app.py:
```diff
@@ -0,0 +1,33 @@
+
+
+def init_app() -> FlaskApp:
+	"""Initialize the Foca app."""
```
FOCA, not Foca

uniqueg commented on tesk/config.yaml:
```diff
@@ -0,0 +1,82 @@
+  specs:
+    - path: tesk/api/task_execution_service.openapi.yaml
+      add_operation_fields:
+        x-openapi-router-controller: tesk.ga4gh.tes.server
```
Don't forget to adapt this when changing the directory structure of the app (as per #156) or the name of the controller module

uniqueg commented on tesk/ga4gh/tes/server.py:
See discussion in #156 to consider renaming this file or putting its contents in the package-level under `__init__.py`.

uniqueg commented on tesk/ga4gh/tes/service_info.py:
I would leave these out for now, as they are not used.

uniqueg commented on tesk/ga4gh/tes/task_runs.py:
As above

uniqueg commented on tests/test_unit/test_services/test_filer.py:
```diff
@@ -4,6 +4,7 @@
 from io import StringIO
 from unittest.mock import patch
 
+from assertThrows import AssertThrowsMixin
```
I think this is a local import, no? In that case, it should be part of the third block of imports.

JaeAeich replied:
Imports are organised using [isort](https://pycqa.github.io/isort/) rules.

uniqueg commented on .pre-commit-config.yaml:
Please remove, as it's not part of this PR.

JaeAeich replied:
:sob:

uniqueg commented on deployment/config.yaml:
```diff
@@ -0,0 +1,82 @@
+    status_member: ['status']
+    exceptions: tesk.exceptions.exceptions
+
+#tesk:
+#  service_list:
+
```
Please remove - we can do this later. Also, probably this should then use FOCA's custom config handler (together with a Pydantic schema for the custom config)

uniqueg commented on tesk/api/specs/task_execution_service.openapi.yaml:
Please include a release tag or commit hash in the filename. Just so that we are sure about the _exact_ version (because there are a lot of commits that have `1.1.0` in `info.version` inside the specs themselves.

JaeAeich commented:
@uniqueg have you encountered this error.

```
❯ python3 tesk/app.py
Traceback (most recent call last):
  File "/home/jaeaeich/Desktop/TESK/tesk/app.py", line 7, in <module>
    from connexion import FlaskApp
  File "/home/jaeaeich/Desktop/TESK/.venv/lib/python3.11/site-packages/connexion/__init__.py", line 32, in <module>
    from .apps.flask_app import FlaskApp
  File "/home/jaeaeich/Desktop/TESK/.venv/lib/python3.11/site-packages/connexion/apps/flask_app.py", line 151, in <module>
    class FlaskJSONEncoder(json.JSONEncoder):
                           ^^^^^^^^^^^^^^^^
AttributeError: module 'flask.json' has no attribute 'JSONEncoder'
```

The fix is in the connexion package to import json rather than falsk.json, but obv this is not the solution.
The other is to do update the poetry package connexion, this works but since I dont have connexion as direct dep rather an indirect one via foca idk how to make it stick.

Either way just so that I don;t keep breaking stuff I'll add a CI test to probe the API and see it its working or not, but maybe the solution would be to check this think in FOCA itself.

uniqueg commented on tesk/app.py:
```diff
@@ -0,0 +1,52 @@
+def init_app() -> FlaskApp:
+	"""Initialize the FOCA app."""
+
+	# Use environment variables, this will be useful for docker
+	config_path_env = os.getenv('TESK_FOCA_CONFIG_PATH')
+	config_filename_env = os.getenv('TESK_FOCA_CONFIG_FILENAME', 'config.yaml')
```
I agree. But why two?

uniqueg commented on tesk/app.py:
```diff
@@ -0,0 +1,52 @@
+	return foca.create_app()
+
+
+if __name__ == '__main__':
+	app = init_app()
+	app.run()
+
```
Please remove, it's already at the bottom (where it belongs)

uniqueg commented on tesk/app.py:
```diff
@@ -0,0 +1,52 @@
+	app.run()
+
+
+def main() -> None:
+	"""Run FOCA application."""
+	app = init_app()
+	app.run(port=app.port)
```
Is this actually being used anywhere?

JaeAeich replied:
poetry console script.

in you virtual env do
- `poetry install`
- `api`

uniqueg replied:
But it seems redundant to me. If you wanna keep the `main()` function, why don't you call it in the `if __name__ == "__main__"` trap rather than repeating the code?

uniqueg commented on tesk/app.py:
```diff
@@ -0,0 +1,52 @@
+	Install `TESK` and run the console script `api`.
+	"""
+	my_app = init_app()
+	my_app.run()
```
Should we also add `port=app.port` in the call to `my_app.run()`?

uniqueg commented on .safety-policy.yml:
```diff
@@ -10,15 +10,21 @@ scanning-settings:
-    70612:
-      reason: 'No fix available.'
-      expires: '2025-5-18'
+      62019:
+          reason: 'Foca needs to update werkzeug to >=2.3.8.'
+          expires: '2025-5-27'
+      65213:
+          reason: 'Fix is downgrading the pyopenssl package to 21.0.0, doing that breaks all the test.'
+          expires: '2025-5-18'
+      67599:
+          reason: 'No fix available.'
+          expires: '2025-5-18'
+      70612:
+          reason: 'No fix available.'
+          expires: '2025-5-18'
+      70624:
+          reason: 'No fix available.'
+          expires: '2025-5-27'
```
Let's use a consistent indent in the YAML files. It seems you're using 2 chars everywhere else and then 4 chars here

uniqueg commented on tesk/api/ga4gh/tes/controllers.py:
```diff
@@ -0,0 +1,71 @@
+
+import logging
+
+# from connexion import request  # type: ignore
```
Please remove and add it if/when you actually need it.

JaeAeich replied:
I need controllers because its in specs as the function ID, parsing will fail without it.

JaeAeich commented:
![image](https://github.com/elixir-cloud-aai/TESK/assets/100477031/10c08b95-293f-400d-bedf-f50b0ac8e3aa)
what does this mean?
```
[2024-05-30 15:32:36,244: INFO    ] Registered database 'access_control_db' at URI 'mongodb':'27017' with Flask application. [foca.database.register_mongodb]
[2024-05-30 15:32:36,244: INFO    ] Added database collection 'policy_rules'. [foca.database.register_mongodb]
```

uniqueg commented:
> @uniqueg have you encountered this error.
> 
> ```
> ❯ python3 tesk/app.py
> Traceback (most recent call last):
>   File "/home/jaeaeich/Desktop/TESK/tesk/app.py", line 7, in <module>
>     from connexion import FlaskApp
>   File "/home/jaeaeich/Desktop/TESK/.venv/lib/python3.11/site-packages/connexion/__init__.py", line 32, in <module>
>     from .apps.flask_app import FlaskApp
>   File "/home/jaeaeich/Desktop/TESK/.venv/lib/python3.11/site-packages/connexion/apps/flask_app.py", line 151, in <module>
>     class FlaskJSONEncoder(json.JSONEncoder):
>                            ^^^^^^^^^^^^^^^^
> AttributeError: module 'flask.json' has no attribute 'JSONEncoder'
> ```
> 
> The fix is in the connexion package to import json rather than falsk.json, but obv this is not the solution. The other is to do update the poetry package connexion, this works but since I dont have connexion as direct dep rather an indirect one via foca idk how to make it stick.
> 
> Either way just so that I don;t keep breaking stuff I'll add a CI test to probe the API and see it its working or not, but maybe the solution would be to check this think in FOCA itself.



> @uniqueg have you encountered this error.
> 
> ```
> ❯ python3 tesk/app.py
> Traceback (most recent call last):
>   File "/home/jaeaeich/Desktop/TESK/tesk/app.py", line 7, in <module>
>     from connexion import FlaskApp
>   File "/home/jaeaeich/Desktop/TESK/.venv/lib/python3.11/site-packages/connexion/__init__.py", line 32, in <module>
>     from .apps.flask_app import FlaskApp
>   File "/home/jaeaeich/Desktop/TESK/.venv/lib/python3.11/site-packages/connexion/apps/flask_app.py", line 151, in <module>
>     class FlaskJSONEncoder(json.JSONEncoder):
>                            ^^^^^^^^^^^^^^^^
> AttributeError: module 'flask.json' has no attribute 'JSONEncoder'
> ```
> 
> The fix is in the connexion package to import json rather than falsk.json, but obv this is not the solution. The other is to do update the poetry package connexion, this works but since I dont have connexion as direct dep rather an indirect one via foca idk how to make it stick.
> 
> Either way just so that I don;t keep breaking stuff I'll add a CI test to probe the API and see it its working or not, but maybe the solution would be to check this think in FOCA itself.

No, I haven't seen the error.

You can't use Connexion 3, this will mess with everything. As you know, I'm working on that, but it will still take a while, maybe some weeks. We _could_ update to the latest Connexion 2 in FOCA, which is 2.14.x - if that helps you.

But if it's just about the typing, try `App` instead of `FlaskApp`, or just ignore it if you can. With the upcoming upgrade to Connexion 3, this should be fixed.

JaeAeich commented:
I have fixed it with update, so the version in the current 'poetry.lock' work but just so that they don't break with any accidental {up/down}grades I have added a CI check as well, would have added it later but I think it needed rn with this.

uniqueg commented:
> ![image](https://private-user-images.githubusercontent.com/100477031/335136240-10c08b95-293f-400d-bedf-f50b0ac8e3aa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcwODMwMDcsIm5iZiI6MTcxNzA4MjcwNywicGF0aCI6Ii8xMDA0NzcwMzEvMzM1MTM2MjQwLTEwYzA4Yjk1LTI5M2YtNDAwZC1iZWRmLWY1MGIwYWM4ZTNhYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUzMFQxNTI1MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNzJjMGM2YWI4OTQyYWM1MTY0MmM4NjdhMzIyNWIxNmVhZjdmYTY2NzVkNGIzMTQ1MzJkODRjNmZjZGEwZDc0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.T0HL_xpUbKc75KNifHF4cnVbk7h8LXhc7GTbTcsvAB0) what does this mean?
> 
> ```
> [2024-05-30 15:32:36,244: INFO    ] Registered database 'access_control_db' at URI 'mongodb':'27017' with Flask application. [foca.database.register_mongodb]
> [2024-05-30 15:32:36,244: INFO    ] Added database collection 'policy_rules'. [foca.database.register_mongodb]
> ```



> ![image](https://private-user-images.githubusercontent.com/100477031/335136240-10c08b95-293f-400d-bedf-f50b0ac8e3aa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcwODMwMDcsIm5iZiI6MTcxNzA4MjcwNywicGF0aCI6Ii8xMDA0NzcwMzEvMzM1MTM2MjQwLTEwYzA4Yjk1LTI5M2YtNDAwZC1iZWRmLWY1MGIwYWM4ZTNhYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUzMFQxNTI1MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNzJjMGM2YWI4OTQyYWM1MTY0MmM4NjdhMzIyNWIxNmVhZjdmYTY2NzVkNGIzMTQ1MzJkODRjNmZjZGEwZDc0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.T0HL_xpUbKc75KNifHF4cnVbk7h8LXhc7GTbTcsvAB0) what does this mean?
> 
> ```
> [2024-05-30 15:32:36,244: INFO    ] Registered database 'access_control_db' at URI 'mongodb':'27017' with Flask application. [foca.database.register_mongodb]
> [2024-05-30 15:32:36,244: INFO    ] Added database collection 'policy_rules'. [foca.database.register_mongodb]
> ```

This is the access control configuration that is automatically set up - this was a huge PR that was going on for a year, and we eventually merged it even though it was still very messy. I will refactor it along the way, such that it doesn't show up if it's not explicitly activated (plus many other necessary improvements). It doesn't make sense to just do that stuff when nobody is asking for it, especially when you didn't even set up a database for handling the permissions.

However, given that it's lazily creating the database connection, as long as you don't try to do any actual DB operations, this shouldn't interfere with anything. So I guess you can leave it for now - unless you actually _want_ to set access control, maybe together with Rahul. You need to do it at some point anyway, and IF it's set up, you could implement 403 in your controllers without revisiting them again later, so it does make some sense to figure this out first.

What speaks against implementing this now, however, is that - as I said - the implementation is still very messy and will very likely undergo considerable changes when I refactor it. It shouldn't change the access policies etc. though, mostly just the configuration and internal implementation in FOCA, so I don't expect that changes on your end would be considerable once its cleaned up in a future FOCA release.

Up to you...

uniqueg commented:
> I have fixed it with update, so the version in the current 'poetry.lock' work but just so that they don't break with any accidental {up/down}grades I have added a CI check as well, would have added it later but I think it needed rn with this.



> I have fixed it with update, so the version in the current 'poetry.lock' work but just so that they don't break with any accidental {up/down}grades I have added a CI check as well, would have added it later but I think it needed rn with this.

Suits me fine. No need to create a release just for that then, before the Connexion 3 migration.

uniqueg commented:
> Add `foca` configs.
> 
> fixes - #156

I think you could improve on that description a little :upside_down_face: 

Something like:

- Add [FOCA](https://github.com/elixir-cloud-aai/foca) archetype configuration for API server
- Add dummy API controllers
- API server starts up with (non-functional) Swagger UI at `http://127.0.0.1:8080/ga4gh/tes/v1/ui`; API specs hosted at `http://127.0.0.1:8080/ga4gh/tes/v1/openapi.json`

uniqueg commented on tesk/app.py:
```diff
@@ -0,0 +1,53 @@
+	app.run(port=app.port)
+
+
+my_app = init_app()
+my_app.run(port=my_app.port)
```
This code should definitely be inside an `if __name__ === "__main__"` trap!

Or, rather, do:

```py
if __name__ == "__main__":
  main()
```

uniqueg commented:
Awesome!

I changed the PR title to `feat` because I think it's worth it :)

Merging now
